components/gitpod-db/src/user-to-team-migration-service.ts

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { AdditionalUserData, Team, User } from "@gitpod/gitpod-protocol";
+import { ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { inject, injectable } from "inversify";
+import { ProjectDB } from "./project-db";
+import { TeamDB } from "./team-db";
+import { ResponseError } from "vscode-jsonrpc";
+import { WorkspaceDB } from "./workspace-db";
+import { TypeORM } from "./typeorm/typeorm";
+import { log, LogContext } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { UserDB } from "./user-db";
+
+@injectable()
+export class UserToTeamMigrationService {
+    @inject(TeamDB) protected readonly teamDB: TeamDB;
+    @inject(UserDB) protected readonly userDB: UserDB;
+    @inject(ProjectDB) protected readonly projectDB: ProjectDB;
+    @inject(WorkspaceDB) protected readonly workspaceDB: WorkspaceDB;
+    @inject(TypeORM) protected readonly typeorm: TypeORM;
+
+    async migrateUser(user: User): Promise<User> {
+        if (!(await this.needsMigration(user))) {
+            return user;
+        }
+        AdditionalUserData.set(user, { isMigratedToTeamOnlyAttribution: true });
+        await this.userDB.storeUser(user);
```

Comment:
For my own curiosity, why do we flag the user as `isMigratedToTeamOnlyAttribution` and save it (if that's what this does) before we've completed the migration (which I assume is happening in `internalMigrateUser()`?

Reply:
Because I wanted to reduce the likelihood of running migrations onnthe same user concurrently